### DISCUSSION TITLE
Medplum Cluster Upgrade Process

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#2778] [General]
# Medplum Cluster Upgrade Process

## tl;dr - what you need to know as a super admin

At present, upgrading a Medplum cluster is somewhat adhoc.  You probably run `medplum aws update-server` to get new features, but not much beyond that.  You might be aware of the [Super Admin](https://www.medplum.com/docs/self-hosting/super-admin-guide) guide, but probably only touch it if something is broken.

There are a few big projects on the horizon that will be more involved.  Some of these will require migrations that may take upwards of hours to complete.  It is our goal to make this as simple, straightforward, and reliable as possible. (See "Examples of upcoming data migrations" below)

The main point of this discussion is the design of a new CLI command `medplum aws update`, which will orchestrate CDK upgrades, schema upgrades, and data migrations.

## Update 2023-12-13

This work is underway.

1. Data Migrations are implemented, tracked, and included in the version system logic.  Every data migration will trigger a minor version bump (i.e., the `y` in `x.y.z`)
2. The first version of `medplum aws update` is in beta testing.  It retrieves the list of necessary version deployments, iteratively deploys and runs data migrations.

## Operating Principles

1. Medplum server should always be in working state
2. Medplum server should never require planned downtime
3. Migrations can be slow, but they should be methodical and robust
4. Migrations should be as automatic as possible 

## Definitions

1. "Infra migration" - changes to AWS infrastructure, written in AWS CDK, executed by AWS CloudFormation
2. "Schema migration" - changes to Postgres database schema, executed by Medplum server at server startup
3. "Data migration" - bulk updates to data in Medplum server, currently done one-off with super admin tools

## Current State

1. Self-hosting is functional, but not as robust as it should be
    a. Use Medplum CLI to update server and app, but very simple
2. Infra is often setup once, and not updated
    a. AWS CDK can update infra, but we have not trained users to do it regularly
3. Server runs "schema migrations", but no "data migrations"
4. No guarantees about the sequencing of server upgrades
5. No way to enforce "you need to run this infra upgrade task before upgrading to version X"
6. No way to enforce "you need to run this data migration task before upgrading to version X"

## Proposal

1. Add "data migrations", separate from "schema migrations"
    a. "Schema migration" - run raw SQL to alter table definitions, should happen fast (under 1 min)
    b. "Data migrations" - use `Repository` to update data or reindex data, can be slow (~hours)
2. Define convention that "major" and "minor" version bumps require data migration step
    a. At present, "minor" version bumps do not have any particular meaning in Medplum
    b. "patch" version bumps are common, usually new features and bug fixes, and do not have any migration requirements
    c. "major" version bumps are rare
    d. "minor" version bumps are now intended to represent a release that requires server maintenance
3. Add Medplum CLI tool to handle all upgrades
    a. New command `medplum aws update [version]`
    b. Replaces `medplum aws update-server` and `medplum aws update-app`
    c. Determines sequence of intermediate versions required to fully upgrade the cluster
    d. For each required version:
        1. Deploy necessary CDK upgrades (`cdk deploy`)
        2. Deploy server versions (which automatically run schema migrations) (`medplum aws update-server vX.Y.Z`)
        3. Run data migrations
    e. Can run in a loop to execute the specific sequence that is required
4. Medplum server will enforce necessary requirements at server startup
    a. If CDK not upgraded to required version, then server will not start
    b. If data migrations not run, then server will not start

## Data Migrations

1. Similar to schema migrations, data migrations will live in a folder in `packages/server/src`
2. We will add a new super admin endpoint such as `/admin/super/migrate`
3. When a super admin successfully `POST`s to that endpoint, it will create a new `AsyncJob` that runs all pending data migrations
4. Current state of data migrations will be tracked in a new column in the `DatabaseMigration` table
5. The async job will run all pending migrations sequentially
6. TODO: Need a mechanism of tracking that migrations are currently running
    a. UPDATE 2023-12-13 - This is now managed in `AsyncJob` resources
7. TODO: Need a mechanism to guarantee that only one data migration job can run at a time

## Examples of upcoming data migrations

1. Project linking (i.e., "R4" project)
    a. Need to move all "R4" resources into the new "R4" project
    b. Requires updating all StructureDefinition, SearchParameter, ValueSet, and CodeSystem resources
2. Client assigned IDs
    a. Need to add new `v5uuid` column to all resource tables
    b. Need to compute, set, and index
3. Terminology service
    a. Need to update `ValueSetElements` to be project-aware
4. References update
    a. Need to add new `X_Reference` tables for all resource types
    b. Need to reindex all resources with references (basically everything)

**Comment by rahul1** [comment]
Great writeup! 

How would data migrations work on multi-tenant instances? Would all projects have to be upgraded at the same time, or could projects opt-in individually. 

If the former, we should probably define some level of contract about what will/won't be considered a "data migration". For example, could the content for core FHIR resources be modified, or only content related to Medplum defined resources?

**Reply by codyebberson** [reply]
This is a good question.  In my opinion, it's out of scope for now, but it's worth considering.

One idea that we discussed offline, which I did not include in the writeup, is the idea that every resource table could have a `dataVersion` column.  We discussed it in the context of guaranteeing exact upgrades (i.e., how to avoid pagination issues in large upgrades), and how to recover from errors in the middle of bad upgrades.  But that same `dataVersion` column could be used for tracking resources on a per-project basis.

My intuition is that there is a clean divide between features that are eligible for per-project or not.  All of the examples listed (project linking, client assigned IDs, etc) are really internal implementation details, and we would not want to carry around support for multiple data versions for any longer than we have to.

An interesting counterexample is R4-to-R5, in which case we will definitely need to maintain some notion of "primary version", and it will likely be per-project.  In that case, we will probably need a totally parallel set of R5 database tables, so maybe that's not a good example.

**Comment by codyebberson** [comment]
Breaking this down into actionable next steps:

- [ ] Add new folder for data migrations
- [ ] Add new super admin endpoint to execute pending data migrations
    - [ ] Create `AsyncJob` for running data migrations
- [ ] Update `publish.sh` to detect data migrations and use minor version bump
- [ ] Add new CLI command for `medplum aws update`
    - [ ] Determine which versions will be required
    - [ ] Run CDK updates
    - [ ] Run server updates
    - [ ] Wait for server update to complete
    - [ ] Run data migrations
    - [ ] Wait for data migrations to complete
    - [ ] Loop if necessary

Tracking tickets:
- https://github.com/medplum/medplum/issues/2784
- https://github.com/medplum/medplum/issues/2785
